lighthouse-core/fraggle-rock/gather/navigation-runner.js

```diff
@@ -5,116 +5,220 @@
-  await driver._page.goto(requestedUrl, {waitUntil: ['load', 'networkidle2']});
+  const {driver, config, requestedUrl} = navigationContext;
+
+  try {
+    const {finalUrl, warnings} = await gotoURL(driver, requestedUrl, {
+      ...navigationContext.navigation,
+      maxWaitForFcp: config.settings.maxWaitForFcp,
+      maxWaitForLoad: config.settings.maxWaitForLoad,
+      waitUntil: navigationContext.navigation.pauseAfterFcpMs ? ['fcp', 'load'] : ['load'],
+    });
 
-  // TODO(FR-COMPAT): disable all throttling settings
-  // TODO(FR-COMPAT): capture page load errors
+    return {finalUrl, navigationError: undefined, warnings};
+  } catch (err) {
+    if (!(err instanceof LighthouseError)) throw err;
+    if (err.code !== 'NO_FCP' && err.code !== 'PAGE_HUNG') throw err;
+    return {finalUrl: requestedUrl, navigationError: err, warnings: []};
+  }
 }
 
+/**
+ * @param {NavigationContext} navigationContext
+ * @param {PhaseState} phaseState
+ */
+async function _collectNetworkRecords(navigationContext, phaseState) {
+  const devtoolsLogArtifactDefn = phaseState.artifactDefinitions.find(
+    definition => definition.gatherer.instance.meta.symbol === DevtoolsLog.symbol
+  );
+  if (!devtoolsLogArtifactDefn) return undefined;
+
+  const devtoolsLogArtifactId = devtoolsLogArtifactDefn.id;
+  const artifactDefinitions = [devtoolsLogArtifactDefn];
+  await collectPhaseArtifacts({...phaseState, phase: 'getArtifact', artifactDefinitions});
+
+  const devtoolsLog = await phaseState.artifactState.getArtifact[devtoolsLogArtifactId];
+  const records = await NetworkRecords.request(devtoolsLog, navigationContext);
+  return records;
+}
+
+/**
+ * @param {NavigationContext} navigationContext
+ * @param {PhaseState} phaseState
+ * @param {UnPromise<ReturnType<typeof _setupNavigation>>} setupResult
+ * @param {UnPromise<ReturnType<typeof _navigate>>} navigateResult
+ * @return {Promise<{artifacts: Partial<LH.GathererArtifacts>, warnings: Array<LH.IcuMessage>, pageLoadError: LH.LighthouseError | undefined}>}
+ */
+async function _computeNavigationResult(
+  navigationContext,
+  phaseState,
+  setupResult,
+  navigateResult
+) {
+  const {navigationError, finalUrl} = navigateResult;
+  const warnings = [...setupResult.warnings, ...navigateResult.warnings];
+  const networkRecords = await _collectNetworkRecords(navigationContext, phaseState);
+  const pageLoadError = networkRecords
+    ? getPageLoadError(navigationError, {
+      url: finalUrl,
+      loadFailureMode: navigationContext.navigation.loadFailureMode,
+      networkRecords,
+    })
+    : navigationError;
+
+  if (pageLoadError) {
+    const locale = navigationContext.config.settings.locale;
+    const localizedMessage = i18n.getFormatted(pageLoadError.friendlyMessage, locale);
+    log.error('NavigationRunner', localizedMessage, navigationContext.requestedUrl);
+
+    return {artifacts: {}, warnings: [...warnings, pageLoadError.friendlyMessage], pageLoadError};
+  } else {
```

Comment:
nit: don't need `else`

Reply:
ack

I intentionally added the `else` to make explicit the two paths this can take. IMO, without the `else`  it's very easy to miss the `return` and think that the `else` bit is executed even in the case of a `pageLoadError`. I'm all for the early return pattern to avoid nesting but when it's a one-line and/or early check where the intention is clearer :) Convinced? or still prefer no `else`?

Reply:
I still prefer no `else` but I won't die on this hill, it's your choice.